File: src/core/thumbnail.py
"""Thumbnail generation with disk cache."""

import hashlib
import sys
from pathlib import Path

from PIL import Image

from .image_loader import load_image

THUMBNAIL_SIZE = 256


def _cache_key(path: Path) -> str:
    """Generate a cache key from absolute path + mtime."""
    path = Path(path).resolve()
    mtime = path.stat().st_mtime
    raw = f"{path}:{mtime}"
    return hashlib.sha256(raw.encode()).hexdigest()


class ThumbnailCache:
    def __init__(self, cache_dir: Path | None = None):
        if cache_dir is None:
            cache_dir = Path.home() / ".cache" / "omarchy-imageview"
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)

    def get_thumbnail(self, path: Path) -> Image.Image | None:
        """Get a thumbnail for the given image path. Uses cache if available."""
        path = Path(path)
        if not path.is_file():
            return None

        try:
            key = _cache_key(path)
        except OSError:
            return None

        cached = self.cache_dir / f"{key}.png"
        if cached.exists():
            try:
                return Image.open(cached)
            except Exception:
                cached.unlink(missing_ok=True)

        return self._generate(path, cached)

    def _generate(self, path: Path, cache_path: Path) -> Image.Image | None:
        """Generate thumbnail, save to cache, return it."""
        img = load_image(path)
        if img is None:
            return None

        try:
            img.thumbnail((THUMBNAIL_SIZE, THUMBNAIL_SIZE), Image.LANCZOS)
            img.save(cache_path, "PNG")
            return img
        except Exception as e:
            print(f"Thumbnail generation failed for {path}: {e}", file=sys.stderr)
            return None

File: src/core/thumbnail.rs
use image::DynamicImage;
use sha2::{Sha256, Digest};
use std::fs;
use std::path::{Path, PathBuf};

const THUMBNAIL_SIZE: u32 = 256;

pub fn cache_key(path: &Path) -> Option<String> {
    let abs = path.canonicalize().ok()?;
    let meta = fs::metadata(&abs).ok()?;
    let mtime = meta.modified().ok()?;
    let raw = format!("v2:{}:{:?}", abs.display(), mtime);
    let hash = Sha256::digest(raw.as_bytes());
    Some(format!("{:x}", hash))
}

pub struct ThumbnailCache {
    cache_dir: PathBuf,
}

impl ThumbnailCache {
    pub fn new(cache_dir: Option<PathBuf>) -> Self {
        let dir = cache_dir.unwrap_or_else(|| {
            dirs::cache_dir()
                .unwrap_or_else(|| PathBuf::from("/tmp"))
                .join("omarchy-imageview")
        });
        fs::create_dir_all(&dir).ok();
        Self { cache_dir: dir }
    }

    pub fn get_thumbnail(&self, path: &Path) -> Option<DynamicImage> {
        if !path.is_file() {
            return None;
        }
        let key = cache_key(path)?;
        let cached = self.cache_dir.join(format!("{}.png", key));

        if cached.exists() {
            if let Ok(img) = image::open(&cached) {
                return Some(img);
            }
            fs::remove_file(&cached).ok();
        }

        self.generate(path, &cached)
    }

    fn generate(&self, path: &Path, cache_path: &Path) -> Option<DynamicImage> {
        let img = super::image_loader::load_image(path)?;
        let thumb = img.thumbnail(THUMBNAIL_SIZE, THUMBNAIL_SIZE);
        thumb.save(cache_path).ok()?;
        Some(thumb)
    }
}
